src/e2e-browser/acp.testcafe.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { ClientFunction } from "testcafe";
+import { config } from "dotenv-flow";
+import { resolve } from "path";
+import { essUser, essUserLogin } from "./roles";
+import type { getHelpers } from "../../.codesandbox/sandbox/src/end-to-end-test-helpers";
+
+// E2eHelpers is a global defined in .codesandbox/sandbox/src/end-to-end-helper.
+// Since code inside of ClientFunction is executed in that context in the browser,
+// that variable is available to it - but as far as TypeScript is concerned,
+// it is executed in the context of this test file.
+// Hence, we just declare this variable to be of the  type here.
```

Comment:
Missing the actual type here (although I tend to avoid putting actual specifics in comments if possible, so I'd just allude to the nature of the type (in English) in the comment, not the actual type name itself):
```
// Hence, we just declare this variable to be of the ????? type here.
```

Reply:
The type is `ReturnType<typeof getHelpers>` - don't know why I inserted a double space there. Possibly I inteded to type "to be of the same type" - would that help?

Reply:
Yeah sure, that's more readable alright. To be honest though, I still don't understand the overall comment, but I certainly appreciate the obvious effort you put into writing it (and the very extensive README). I need to read up on TestCafe's `ClientFunction` thing I guess, and also on CodeSandbox, and then carefully re-read through this whole PR, as I suspect there's lots of cool stuff going on here! Cheers.

Reply:
Yeah it's all a bit hairy. Maybe as a helpful metaphor, imagine that the test code was written in Java. The code inside `ClientFunction` would then still be JavaScript, because it's not actually test code.

The difference is that in Java, the JS would have to be written inside a string, which the test runner would then inject into the page - and then of course transform the return value back into data that Java understands. Since both are in JS, however, TestCafe can take the AST of the contents of `ClientFunction`, serialise that to inject it into the page, and then inject the resulting value into the test again.